File: services/prediction-service/src/prediction_service/store.py
from typing import Protocol

import psycopg

from prediction_service.models import ETAPredictedEvent


class ProcessedEventStore(Protocol):
    def is_processed(self, event_id: str) -> bool: ...

    def mark_processed_with_outbox_event(
        self,
        event_id: str,
        outbox_event: ETAPredictedEvent,
    ) -> None: ...


class OutboxStore(Protocol):
    def get_pending_events(self) -> list[ETAPredictedEvent]: ...

    def mark_published(self, event_id: str) -> None: ...


class PostgresPredictionStore:
    def __init__(self, database_url: str) -> None:
        self._database_url = database_url

    def is_processed(self, event_id: str) -> bool:
        with psycopg.connect(self._database_url) as connection:
            with connection.cursor() as cursor:
                cursor.execute(
                    """
                    SELECT EXISTS (
                        SELECT 1
                        FROM prediction.processed_events
                        WHERE event_id = %s
                    )
                    """,
                    (event_id,),
                )

                row = cursor.fetchone()

        return bool(row[0])

    def mark_processed_with_outbox_event(
        self,
        event_id: str,
        outbox_event: ETAPredictedEvent,
    ) -> None:
        with psycopg.connect(self._database_url) as connection:
            with connection.cursor() as cursor:
                # Persist the processed event and its outgoing event in the same transaction
                # so either both records are committed or neither is.
                cursor.execute(
                    """
                    INSERT INTO prediction.processed_events (event_id)
                    VALUES (%s)
                    """,
                    (event_id,),
                )

                cursor.execute(
                    """
                    INSERT INTO prediction.outbox_events (
                        event_id,
                        event_type,                        
                        payload
                    )
                    VALUES (%s, %s, %s)
                    """,
                    (
                        outbox_event.event_id,
                        outbox_event.event_type,
                        outbox_event.model_dump_json(),
                    ),
                )

    def get_pending_events(self) -> list[ETAPredictedEvent]:
        with psycopg.connect(self._database_url) as connection:
            with connection.cursor() as cursor:
                cursor.execute(
                    """
                    SELECT payload
                    FROM prediction.outbox_events
                    WHERE published_at IS NULL
                    ORDER BY created_at
                    """
                )

                rows = cursor.fetchall()

        return [ETAPredictedEvent.model_validate(row[0]) for row in rows]

    def mark_published(self, event_id: str) -> None:
        with psycopg.connect(self._database_url) as connection:
            with connection.cursor() as cursor:
                cursor.execute(
                    """
                    UPDATE prediction.outbox_events
                    SET published_at = NOW()
                    WHERE event_id = %s
                    """,
                    (event_id,),
                )

File: services/routing-service/store.go
package main

import (
	"database/sql"
	"encoding/json"
)

// OutboxStore abstracts pending event retrieval and publication tracking,
// keeping the outbox publishing flow independent from PostgreSQL.
type OutboxStore interface {
	GetPendingEvents() ([]RouteCalculatedEvent, error)
	MarkPublished(eventID string) error
}

// ProcessedEventStore keeps idempotency state outside the processing logic,
// so duplicate event detection can use either in-memory or persistent storage.
type ProcessedEventStore interface {
	IsProcessed(eventID string) (bool, error)
	MarkProcessedWithOutboxEvent(
		eventID string,
		outboxEvent RouteCalculatedEvent,
	) error
}

type InMemoryProcessedEventStore struct {
	processed map[string]struct{}
}

func NewInMemoryProcessedEventStore() *InMemoryProcessedEventStore {
	return &InMemoryProcessedEventStore{
		processed: make(map[string]struct{}),
	}
}

func (s *InMemoryProcessedEventStore) IsProcessed(eventID string) (bool, error) {
	_, exists := s.processed[eventID]
	return exists, nil
}

func (s *InMemoryProcessedEventStore) MarkProcessedWithOutboxEvent(
	eventID string,
	outboxEvent RouteCalculatedEvent,
) error {
	s.processed[eventID] = struct{}{}
	return nil
}

// PostgresRoutingStore persists Routing Service reliability state,
// including processed event IDs and transactional outbox events.
type PostgresRoutingStore struct {
	db *sql.DB
}

func NewPostgresRoutingStore(db *sql.DB) *PostgresRoutingStore {
	return &PostgresRoutingStore{
		db: db,
	}
}

func (s *PostgresRoutingStore) IsProcessed(eventID string) (bool, error) {
	var exists bool

	err := s.db.QueryRow(
		`
        SELECT EXISTS (
            SELECT 1
            FROM routing.processed_events
            WHERE event_id = $1
        )
        `,
		eventID,
	).Scan(&exists)

	if err != nil {
		return false, err
	}

	return exists, nil
}

func (s *PostgresRoutingStore) MarkProcessedWithOutboxEvent(
	eventID string,
	outboxEvent RouteCalculatedEvent,
) error {
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	if _, err := tx.Exec(
		`
		INSERT INTO routing.processed_events (event_id)
		VALUES ($1)
		`,
		eventID,
	); err != nil {
		return err
	}

	payload, err := json.Marshal(outboxEvent)
	if err != nil {
		return err
	}

	if _, err := tx.Exec(
		`
		INSERT INTO routing.outbox_events (
			event_id,
			event_type,
			payload
		)
		VALUES ($1, $2, $3)
		`,
		outboxEvent.EventID,
		outboxEvent.EventType,
		payload,
	); err != nil {
		return err
	}

	return tx.Commit()
}

func (s *PostgresRoutingStore) GetPendingEvents() ([]RouteCalculatedEvent, error) {
	rows, err := s.db.Query(`
		SELECT payload
		FROM routing.outbox_events
		WHERE published_at IS NULL
		ORDER BY created_at
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var events []RouteCalculatedEvent

	for rows.Next() {
		var payload []byte

		if err := rows.Scan(&payload); err != nil {
			return nil, err
		}

		var event RouteCalculatedEvent

		if err := json.Unmarshal(payload, &event); err != nil {
			return nil, err
		}

		events = append(events, event)
	}

	if err := rows.Err(); err != nil {
		return nil, err
	}

	return events, nil
}

func (s *PostgresRoutingStore) MarkPublished(eventID string) error {
	_, err := s.db.Exec(
		`
		UPDATE routing.outbox_events
		SET published_at = NOW()
		WHERE event_id = $1
		`,
		eventID,
	)

	return err
}
